android: implement biometric authentication (fingerprint)

f321x opened the pull request:
Allows to unlock the android app with the android biometric api (e.g. fingerprint). Can be enabled in the settings.
The initial version was LLM generated but i reviewed it and rewrote a couple things.

TODO: checkbox to enable it in the wizard during wallet setup once one password is enforced for creation of new wallets

closes https://github.com/spesmilo/electrum/issues/7560

f321x committed:
android: implement biometric authentication

Allows to unlock the android app with the android biometric api (e.g.
fingerprint). Can be enabled in the settings.

SomberNight commented on electrum/simple_config.py:
```diff
@@ -676,6 +676,11 @@ def __setattr__(self, name, value):
+    WALLET_ANDROID_USE_BIOMETRIC_AUTHENTICATION = ConfigVar('android_use_biometrics', default=False, type_=bool)
+    WALLET_ANDROID_BIOMETRIC_DATA = ConfigVar(
+        'android_biometric_data', default='', type_=str,
+        short_desc=lambda: 'Wallet password encrypted with android biometric authentication key'
```
i18n localization missing

SomberNight reviewed (commented):
just some quick comments, I haven't really looked at this

f321x committed:
qeqrscanner: check requestCode on activity result

ecdsa commented:
this looks good to me, although I did not build an apk,

SomberNight commented on electrum/gui/qml/java_classes/org/electrum/biometry/BiometricActivity.java:
```diff
@@ -0,0 +1,168 @@
+        KeyStore keyStore = KeyStore.getInstance("AndroidKeyStore");
+        keyStore.load(null);
+        return (SecretKey) keyStore.getKey(KEY_NAME, null);
+    }
```
Can another android application get our secret key?

Do we get protection automatically by using [`AndroidKeyStore`](https://developer.android.com/privacy-and-security/keystore#WhichShouldIUse)?

f321x replied:
Using `AndroidKeyStore` prevents other apps from using our secret by assigning it to the Electrum apps specific UID. 
So unless the phone is rooted or compromised it shouldn't be possible for other apps to request any cryptographic operation on the Electrum AndroidKeyStore entry. And each operation requires user authentication as the key has been generated with `setUserAuthenticationRequired(true)`

> Use the Android Keystore provider to let an individual app store its own credentials, which only that app can access. - https://developer.android.com/privacy-and-security/keystore#WhichShouldIUse

> Access control
Android Keystore provides a central component for the storage and use of hardware-backed cryptographic keys, both for apps and for other system components. As such, access to any individual key is normally limited to the app or system component that created the key.

> Keystore domains
> To support this access control, keys are identified to Keystore with a key descriptor. This key descriptor indicates a domain that the descriptor belongs to, together with an identity within that domain.
> Android apps access Keystore using the standard Java Cryptography Architecture, which identifies keys with a string alias. This method of identification maps to the Keystore APP domain internally; the UID of the caller is also included to disambiguate keys from different apps, preventing one app from accessing another's keys.
> Internally, the frameworks code also receives a unique numeric key ID after a key has been loaded. This numeric ID is used as the identifier for key descriptors within the KEY_ID domain. However, access control is still performed: even if one app discovers a key ID for another app's key, it can't use it in normal circumstances. - https://source.android.com/docs/security/features/keystore#access-control

This is the source code of the keystore, you can grep for `caller_uid` to find the permission checks:
https://android.googlesource.com/platform/system/security/+/main/keystore2/src/database.rs

SomberNight commented on electrum/gui/qml/java_classes/org/electrum/biometry/BiometricActivity.java:
```diff
@@ -0,0 +1,168 @@
+                String data = intent.getStringExtra("data"); // Wallet password to encrypt
+                byte[] encrypted = cipher.doFinal(data.getBytes(Charset.forName("UTF-8")));
+                resultIntent.putExtra("data", Base64.encodeToString(encrypted, Base64.NO_WRAP));
+                resultIntent.putExtra("iv", Base64.encodeToString(cipher.getIV(), Base64.NO_WRAP));
```
To be clear, I am not trying to imply that these are the wrong choices, I just want to make sure we understand what we are doing.

f321x replied:
> Crossing into Java (using an intent) is cross-process though

https://developer.android.com/guide/components/processes-and-threads says:

> By default, all components of the same application run in the same process and thread, called the main 

Is this a python for android specific thing to start java activities in a separate process? If i understand the docs correctly "all components" should run in the same process unless explicitly specified to run in another process with an android:process attribute, which i don't find us specifying anywhere.

> does this cipher.doFinal() call out to that external process then?

I think so, yes. Looking at https://android.googlesource.com/platform/frameworks/base/+/508e665/keystore/java/android/security/keystore/AndroidKeyStoreCipherSpiBase.java you can see that they send the data to the keystore through some IPC handler `mMainDataStreamer`. Grep `createMainDataStreamer`, `mMainDataStreamer.doFinal`

SomberNight replied:
>> Crossing into Java (using an intent) is cross-process though
>
> Is this a python for android specific thing to start java activities in a separate process?

Our python/qml/java code is all running in the same process. (What I don't know is whether intent resolution is done inside this same process.)

```
$ adb logcat | grep -F "`adb shell ps | grep org.electrum.electrum | cut -c14-19`"
01-11 06:07:10.530 18591 18591 W zygote64: Activation of cgroup controller +memory failed in path /sys/fs/cgroup/apps/uid_1110227: Invalid argument
01-11 06:07:10.533 18591 18591 I libc    : SetHeapTaggingLevel: tag level set to 0
01-11 06:07:10.537 18591 18591 I Zygote  : Process 18591 created for org.electrum.electrum
[...]
01-11 06:07:10.792 18591 18591 V PythonActivity: My onCreate running
01-11 06:07:10.792 18591 18591 V PythonActivity: loading native lib
[...]
01-11 06:07:11.087 18591 18615 I python  : Initializing Python for Android
[...]
01-11 06:07:11.895 18591 18615 I python  :   0.69 | I | logging | Electrum version: 4.6.2 - https://electrum.org - https://github.com/spesmilo/electrum
```
The two columns after the timestamp are the process id and the thread id.
- `01-11 06:07:10.792 18591 18591 V PythonActivity: My onCreate running`
    - https://github.com/spesmilo/python-for-android/blob/a01269f7799587ad74ee40e0b642d917b8db7d4e/pythonforandroid/bootstraps/qt6/build/src/main/java/org/kivy/android/PythonActivity.java#L100
- `01-11 06:07:10.792 18591 18591 V PythonActivity: loading native lib`
    - https://github.com/spesmilo/python-for-android/blob/a01269f7799587ad74ee40e0b642d917b8db7d4e/pythonforandroid/bootstraps/qt6/build/src/main/java/org/kivy/android/PythonActivity.java#L86-L88
- `01-11 06:07:11.087 18591 18615 I python  : Initializing Python for Android`
    - https://github.com/spesmilo/python-for-android/blob/a01269f7799587ad74ee40e0b642d917b8db7d4e/pythonforandroid/bootstraps/common/build/jni/application/src/Android.mk#L5
    - https://github.com/spesmilo/python-for-android/blob/a01269f7799587ad74ee40e0b642d917b8db7d4e/pythonforandroid/bootstraps/common/build/jni/application/src/Android.mk#L12-L13
    - https://github.com/spesmilo/python-for-android/blob/a01269f7799587ad74ee40e0b642d917b8db7d4e/pythonforandroid/bootstraps/common/build/jni/application/src/start.c#L83
    - https://github.com/spesmilo/python-for-android/blob/a01269f7799587ad74ee40e0b642d917b8db7d4e/pythonforandroid/bootstraps/common/build/jni/application/src/start.c#L309-L311
- `01-11 06:07:11.895 18591 18615 I python  :   0.69 | I | logging | Electrum version: 4.6.2 - https://electrum.org - https://github.com/spesmilo/electrum`
    - https://github.com/spesmilo/electrum/blob/45e1b0da8d3461234db213cf255e5a1f49ff54c0/electrum/logging.py#L336

-----

As intents can be used to communicate between different apps, I was thinking the intent goes out to an external process to decide what to do. But maybe that only happens for implicit intents, and perhaps explicit intents are handled in-process. Not sure.

f321x replied:
Implemented this in https://github.com/spesmilo/electrum/pull/10340/commits/d4c2123db6dccdb0c53995f58e4effe09874fad9

SomberNight replied:
Great. Thanks. I think this is a good approach.

f321x commented on electrum/gui/qml/qebiometrics.py:
```diff
@@ -0,0 +1,133 @@
+        self._start_activity(BiometricAction.DECRYPT, data=encrypted_bundle)
+
+    def _start_activity(self, action: BiometricAction, data: str):
+        self._current_action = action
```
Yeah good idea, doing this now in https://github.com/spesmilo/electrum/compare/0cf8b3ca8c9ca831af5dc65b02af49aad284ef0a..b0800d62232376ff61717c01e23f1a70189fd5b9

SomberNight commented on electrum/gui/qml/qebiometrics.py:
```diff
@@ -0,0 +1,133 @@
+            raise ValueError(f"unsupported {action=}")
+
+        activity.bind(on_activity_result=self._on_activity_result)
+        jPythonActivity.startActivityForResult(intent, self.REQUEST_CODE_BIOMETRIC_ACTIVITY)
```
> As we construct the jIntent with the target class `jBiometricActivity` this is the same as calling setComponent, according to the docs of the Intent constructor you linked. [..]
> So our biometric intent should be safe from hijacking as it explicitly specifies the receiver of the intent, jBiometricActivity.

I was also wondering if another application could force a collision by creating a Java package also named `org.electrum.biometry.BiometricActivity` and how that would be handled.

Is your reading of the docs that by using the `intent = jIntent(jPythonActivity, jBiometricActivity)` constructor (`public Intent (Context packageContext, Class<?> cls)`), `packageContext=jPythonActivity` will get set to something that effectively limits the search to the `org.electrum.electrum` (i.e. the current) android package?

> For another app to be able to launch our intent we would have to allow this by setting android:exported true in the AndroidManifest.xml or register an intent filter which would also set android:exported true.
> https://developer.android.com/guide/topics/manifest/activity-element#exported

Are you assuming the default for `android:exported` is `false`? That doc does not say what the default is :/
Looks like [the default is unclear](https://developer.android.com/privacy-and-security/risks/android-exported)... >.<

SomberNight commented on electrum/gui/qml/qedaemon.py:
```diff
@@ -232,6 +232,12 @@ def load_wallet_task():
+                    if not self._use_single_password and self.daemon.config.WALLET_ANDROID_USE_BIOMETRIC_AUTHENTICATION:
+                        # we need to disable biometric auth if the user creates wallets with different passwords as
+                        # we only store one encrypted password which is not associated to a specific wallet
+                        self._logger.warning(f"biometric authentication disabled, not in single password mode")
```
```suggestion
                        self._logger.warning(f"disabling biometric authentication, not in single password mode")
```

SomberNight reviewed (commented):
Very nice!

f321x committed:
qml: remove pin code authentication

Completely removes the pin code authentication from qml. The config
option in the wallet preferences has been renamed to "Payment
authentication" and now either asks for the Android system
authentication (Biometric or system pin/password) if enabled or will ask
for the wallet password as fallback.

f321x commented on electrum/gui/qml/components/main.qml:
```diff
@@ -771,43 +828,57 @@ ApplicationWindow
+            })
+            dialog.open()
+        } else if (method == 'pin') {
+            var dialog = app.pinDialog.createObject(app, {
```
Implemented this in https://github.com/spesmilo/electrum/pull/10340/commits/23ee523ba70ec361ff401ba2cb5865db77bcc8d6, the pin code / pin dialog logic is fully removed and the password dialog is used as fallback.

Also increases the minimum required Android API version from 29 to 30 to simplify the Java code, otherwise we would need to maintain two different paths a couple things as many API calls have been introduced or changed in 30.

SomberNight commented on electrum/gui/qml/components/Preferences.qml:
```diff
@@ -158,59 +158,83 @@ Pane {
+                                console.log('paymentAuthentication: ' + checked)
+                                if (activeFocus) {
+                                    // will request authentication when checked = false
+                                    Config.paymentAuthentication = !!checked;
```
is the double-negation intended?

f321x commented on electrum/gui/qml/components/WalletDetails.qml:
```diff
@@ -475,6 +475,15 @@ Pane {
+                if (success && Biometrics.isEnabled) {
+                    if (Biometrics.isAvailable) {
+                        // also update the biometric authentication
+                        Biometrics.enable(dialog.password)
```
Okay, yes I see the point, added these changes in https://github.com/spesmilo/electrum/compare/bb8f9772b00d5fcb77d725910b64f279ff0d09b9..9a76759ce330541e6281fb3fb171778585a1c72d.

It will not fall back to the password dialog for signing, sign&broadcast and lightning payments (which are probably the most authentication cases in regular use).

f321x replied:
Okay, yes that's simpler, done this in https://github.com/spesmilo/electrum/compare/9a76759ce330541e6281fb3fb171778585a1c72d..9e42c1215be210ca47428d1d11c38b3d84357090.

> One drawback is that the user can easily forget the wallet password if they have biometrics enabled: they basically never need it.

We could simply prevent the user from falling for this footgun by requiring `wallet_password_only` to disable biometrics. Then they notice "oh what was my password again?" and still have the chance to access the seed phrase or send funds out using the biometrics to do a backup in case they don't have one. Otherwise if they disable biometrics and restart the app they got locked out and cannot re-enable biometrics (only possible after unlock).

SomberNight replied:
> We could simply prevent the user from falling for this footgun by requiring `wallet_password_only` to disable biometrics.

Ok, that sounds reasonable.
If you do that, please add a code comment near that briefly explaining that the auth is there to prevent the user locking themselves out if they forgot the password.

Though note that if the system decides to clear the AndroidKeystore tokens [for whatever reason](https://developer.android.com/reference/android/security/keystore/KeyGenParameterSpec.Builder#setUserAuthenticationRequired(boolean)), and the user forgot the wallet password, they can still get locked out:

> The key will become irreversibly invalidated once the secure lock screen is disabled (reconfigured to None, Swipe or other mode which does not authenticate the user) or when the secure lock screen is forcibly reset (e.g., by a Device Administrator). Additionally, if the key requires that user authentication takes place for every use of the key, it is also irreversibly invalidated once a new biometric is enrolled or once\ no more biometrics are enrolled, unless [setInvalidatedByBiometricEnrollment(boolean)](https://developer.android.com/reference/android/security/keystore/KeyGenParameterSpec.Builder#setInvalidatedByBiometricEnrollment(boolean)) is used to allow validity after enrollment, or KeyProperties.AUTH_DEVICE_CREDENTIAL is specified as part of the parameters to [setUserAuthenticationParameters(int, int)](https://developer.android.com/reference/android/security/keystore/KeyGenParameterSpec.Builder#setUserAuthenticationParameters(int,%20int)).

SomberNight commented on electrum/gui/qml/components/main.qml:
```diff
@@ -761,53 +817,52 @@ ApplicationWindow
+            // wallet has no password
+            qtobject.authProceed()
+            return
+        }
```
I guess this is a pre-existing issue, but shouldn't we call `handleAuthConfirmationOnly` instead?

SomberNight reviewed (approved):
Looks good. Thanks!

PapaMarc commented:
just found a notification on this, and updated to 4.8.0.
Superb gents. kudos
Thanks!
